distribution_metadata.c

```diff
@@ -126,48 +115,36 @@ List *
 LoadShardIntervalList(Oid distributedTableId)
 {
 	List *shardIntervalList = NIL;
-	RangeVar *heapRangeVar = NULL;
-	RangeVar *indexRangeVar = NULL;
-	Relation heapRelation = NULL;
-	Relation indexRelation = NULL;
-	IndexScanDesc indexScanDesc = NULL;
-	const int scanKeyCount = 1;
-	ScanKeyData scanKey[scanKeyCount];
-	HeapTuple heapTuple = NULL;
-
-	heapRangeVar = makeRangeVar(METADATA_SCHEMA_NAME, SHARD_TABLE_NAME, -1);
-	indexRangeVar = makeRangeVar(METADATA_SCHEMA_NAME, SHARD_RELATION_INDEX_NAME, -1);
-
-	heapRelation = relation_openrv(heapRangeVar, AccessShareLock);
-	indexRelation = relation_openrv(indexRangeVar, AccessShareLock);
-
-	ScanKeyInit(&scanKey[0], 1, BTEqualStrategyNumber, F_OIDEQ,
-				ObjectIdGetDatum(distributedTableId));
-
-	indexScanDesc = index_beginscan(heapRelation, indexRelation, SnapshotSelf,
-									scanKeyCount, 0);
-	index_rescan(indexScanDesc, scanKey, scanKeyCount, NULL, 0);
-
-	heapTuple = index_getnext(indexScanDesc, ForwardScanDirection);
-	while (HeapTupleIsValid(heapTuple))
+	Oid argTypes[] = { OIDOID };
+	Datum argValues[] = { ObjectIdGetDatum(distributedTableId) };
+	const int argCount = sizeof(argValues) / sizeof(argValues[0]);
+	int spiStatus = 0;
+	MemoryContext upperContext = CurrentMemoryContext, oldContext = NULL;
+
+	SPI_connect();
+
+	spiStatus = SPI_execute_with_args("SELECT s.id, s.relation_id, s.min_value, "
+									  "s.max_value, p.partition_method, p.key "
+									  "FROM pgs_distribution_metadata.shard AS s "
+									  "JOIN pgs_distribution_metadata.partition AS p "
+									  "ON s.relation_id = p.relation_id "
+									  "WHERE s.relation_id = $1", argCount, argTypes,
+									  argValues, NULL, false, 0);
+	Assert(spiStatus == SPI_OK_SELECT);
+
+	oldContext = MemoryContextSwitchTo(upperContext);
```

Comment:
SPI uses [its own memory context](http://www.postgresql.org/docs/9.4/static/spi-memory.html), which it switches into after connecting and which `SPI_finish` destroys.

If I needed to directly allocate some memory for a return value, I could use `SPI_palloc`, but if I'm _calling_ a function which uses plain-old `palloc`, it's easier to just switch back to the caller's memory context during that call.